cocos/base/CCUserDefault.cpp

```diff
@@ -407,76 +243,71 @@ void UserDefault::setDelegate(UserDefault *delegate)
     _userDefault = delegate;
 }
 
-
-bool UserDefault::isXMLFileExist()
-{
-    return FileUtils::getInstance()->isFileExist(_filePath);
-}
-
-void UserDefault::initXMLFilePath()
+void UserDefault::init()
 {
-    if (! _isFilePathInitialized)
+    if (! _initialized)
     {
-        _filePath += FileUtils::getInstance()->getWritablePath() + XML_FILE_NAME;
-        _isFilePathInitialized = true;
-    }    
-}
-
-// create new xml file
-bool UserDefault::createXMLFile()
-{
-    bool bRet = false;  
+        _filePath = FileUtils::getInstance()->getWritablePath() + USER_DEFAULT_FILENAME;
 
-    pugi::xml_document doc;
-    doc.load_string(R"(<?xml version="1.0" encoding="UTF-8" ?>)"
-"<" USERDEFAULT_ROOT_NAME " />", pugi::parse_full);
+        // construct file mapping
+        _fd = posix_open(_filePath.c_str(), O_WRITE_FLAGS);
+        if (_fd == -1) {
+            log("[Warnning] UserDefault::init open storage file '%s' failed!", _filePath.c_str());
+            return;
+        }
 
-    bRet = doc.save_file(_filePath.c_str(), " ");
+        int filesize = posix_lseek(_fd, 0, SEEK_END);
+        posix_lseek(_fd, 0, SEEK_SET);
 
-    return bRet;
-}
+        if (filesize < _curMapSize) {
+            posix_fsetsize(_fd, _curMapSize);
+        }
+        /// load to memory _values
+        _rwmmap = std::make_shared<mio::mmap_sink>(posix_fd2fh(_fd), 0, _curMapSize);
+        if (_rwmmap->is_mapped()) { // no error
```

Comment:
what if _rwmmap->is_mapped() return false, should have some fallback method here.

Reply:
When the file size is zero, the map will failed,  maybe there need set _curMapSize to 0 when map failed, what do you think?

Reply:
Set _curMapSize to 0 can be ok, but add some fail log is better. Another thing is when you set _curMapSize to 0, next time you flush, you may encounter the error code, better to catch the error_code and print fail log instead. Check the code below:
if (obs.length() > _curMapSize) {
        _rwmmap->unmap();
        std::error_code error;
        _curMapSize <<= 1; // X2
        _rwmmap->map(posix_fd2fh(_fd), 0, _curMapSize, error);
    }

As shown in the flush code, if _curMapSize is 0, I believe map() will failed and result in error_code.

Reply:
Yeah, agree with you.